addons/block_code/serialization/block_name_tree.gd

```diff
@@ -0,0 +1,24 @@
+extends Resource
+
+const BlockNameTreeNode = preload("res://addons/block_code/serialization/block_name_tree_node.gd")
+
+@export var root: BlockNameTreeNode
+@export var canvas_position: Vector2
+
+
+func _init(p_root: BlockNameTreeNode = null, p_canvas_position: Vector2 = Vector2(0, 0)):
+	root = p_root
+	canvas_position = p_canvas_position
```

Comment:
This is an interesting change compared with the current serialization. If I understand correctly:
- The block script has an array of these trees
- There is one of these trees per set of blocks attached together in the canvas
- At the same time the blocks have other blocks as children
- Position in the canvas is set for the trees, because there is no need to save position of blocks attached to others, only to the root block.

Currently this extra tree is not needed:
- The block script has an array of entry blocks
- The blocks have children
- All blocks have position, but the only position that's taken into account is the one for the root blocks

I imagine a user dragging a block that's attached to an empty place in the canvas. In the new way, more data will be changed because a new tree resource will be created (more diff). In the current way, ignored data is persisted but is immediately available.

Not a big deal though, just trying to understand the change.

Reply:
Right, the way the serialization state is done here is with resources. I tried to make it so that the resources contain the bare minimum amount of data to recreate the block trees, but I realize that the resources themselves store quite a bit of data intrinsically as well.

Again I think this might be a good place for iteration. The serialization in this PR is very Godot-like and achieved the goals I wanted (like loading blocks from name and storing minimum data), but it is unfortunate that the resources generate large diffs. I think the `VisualShaderNode`s in the engine could provide some better insight as to the proper way to serialize things like this, so maybe that could be an area of research in the future.

Reply:
I actually like this and think it's an improvement. In the current scenario, if you move a stack of blocks, you'll still get a change because the root block position changed. And there's more in the serialization because every block has a position even if it's actually unused in all but the root block.